### ISSUE TITLE
verify doesn't fail when parameter is mocked

### ISSUE THREAD

**Issue by dalef84** [jonreid/OCMockito#162]
The problem is easier to explain with code.

1. `A` has a `B` property that is always unused
2. When `useC:` is called on `A`, `A` passes `C`'s `string` to `B` (`B` is always `nil`)
3. The test verifies that `B` is called with `C`'s `string`
4. When `C` is not mocked, the test fails as expected
5. When `C` is mocked, the test doesn't fail as expected

```
@import OCHamcrest;
@import OCMockito;
@import XCTest;

@interface C: NSObject
@property (nonatomic, copy) NSString *string;
@end

@implementation C
@end

@interface B: NSObject
- (void)useCString:(NSString *)string;
@end

@implementation B
- (void)useCString:(NSString *)string { }
@end


@interface A: NSObject

- (void)useC:(C *)c;

@property (nonatomic, strong, readonly) B *b; // always uninitialized

@end

@implementation A

- (void)useC:(C *)c
{
    [self.b useCString:c.string];
}

@end

@interface ExampleTests : XCTestCase
@end

@implementation ExampleTests

- (void)test_doesNotMockC_andFailsAsExpected
{
    // Given
    A *a = [[A alloc] init]; // notice I'm not using initWithB
    C *c = [[C alloc] init];

    // When
    [a useC:c];

    // Then
    B *b = mock(B.class); // not used at all
    [verify(b) useCString:c.string];
}

- (void)test_mocksC_doesntFailAsExpected
{
    // Given
    A *a = [[A alloc] init]; // notice I'm not using initWithB
    C *c = mock(C.class);

    // When
    [a useC:c];

    // Then
    B *b = mock(B.class); // not used at all
    [verify(b) useCString:c.string];
}

@end
```

**Comment by dalef84** [comment]
I'm not familiar with this codebase but I hacked on a solution that you can see in [this diff](https://github.com/jonreid/OCMockito/files/6758964/fixdiff.txt). If you think this is a valid approach I could create a PR. 

What I'm doing is I'm trying to keep some relation of `MKTMockingProgress` and the mock it was used on.

**Comment by jonreid** [comment]
Hi @dalef84. Whoo, it's been a long time since I've looked at OCMockito. Just running it again, and debugging through it, was fun. Here's what I've learned so far:

In your code, it seems to be tripping over "what am I verifying?" vs. "what am I stubbing?" This is normally not an issue because it's best to keep those two things cleanly separated in test code. In your example, if I extract `c.string` to a variable so it's outside the `verify` call, the test fails as desired:

    B *b = mock(B.class); // not used at all
    NSString *string = c.string;
    [verify(b) useCString:string];

This may be enough to solve your problem. If not, then we'll have to do something like what you suggest, to give `MKTMockingProgress` some more smarts.

Which way shall we go?

**Comment by dalef84** [comment]
Hi @jonreid ! Thanks for looking into this. I understand that mocking outside of the `verify` call solves the problem but I would suggest we also find a more robust solution since it's very easy to mix these calls and ends up with a false positive. I also thought about enforcing it via eg a linter rule but I couldn't think of any.

**Comment by jonreid** [comment]
Then why don't you open a PR since you have a good idea, and let's go from there.

**Comment by dalef84** [comment]
> Then why don't you open a PR since you have a good idea, and let's go from there.

I tried pulling it together [in this PR](https://github.com/jonreid/OCMockito/pull/164). Let me know what you think 😄

**Comment by jonreid** [comment]
Thanks again!

**Comment by dalef84** [comment]
No problem @jonreid ! Do you think this fix deserves a `6.0.1` tag?

**Comment by jonreid** [comment]
I'm actually planning to add a bit more… If you'd rather have this release sooner, I can just push out 6.0.1. But if you're not in a hurry, it will be in 6.1.0.

**Comment by dalef84** [comment]
Thanks @jonreid waiting for 6.1.0 that sounds fine to me 👍

**Comment by dalef84** [comment]
@jonreid is `v6.1.0` going to be released sometime soon? Still waiting for it 😊

**Comment by jonreid** [comment]
Thank you for the reminder!
Fix now out in v7.0.0
